File: cms-frontend/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface User {
  id: string;
  email: string;
  role: 'ADMIN' | 'EDITOR';
  isActive: boolean;
  createdDate: string;
  updatedDate: string;
  createdBy: string;
  updatedBy: string;
}

export interface UserCreateRequest {
  email: string;
  password: string;
  role: 'ADMIN' | 'EDITOR';
}

export interface UserUpdateRequest {
  email?: string;
  role?: 'ADMIN' | 'EDITOR';
  isActive?: boolean;
}

export interface UserPageResponse {
  content: User[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) {}

  // Get all users with pagination
  getAllUsers(page: number = 0, size: number = 10): Observable<UserPageResponse> {
    const params = new HttpParams()
      .set('pageNumber', page.toString())
      .set('pageSize', size.toString());
    
    return this.http.get<UserPageResponse>(this.apiUrl, { params });
  }

  // Get users by role
  getUsersByRole(role: 'ADMIN' | 'EDITOR', page: number = 0, size: number = 10): Observable<UserPageResponse> {
    const params = new HttpParams()
      .set('pageNumber', page.toString())
      .set('pageSize', size.toString());
    
    return this.http.get<UserPageResponse>(`${this.apiUrl}/role/${role}`, { params });
  }

  // Get active users
  getActiveUsers(page: number = 0, size: number = 10): Observable<UserPageResponse> {
    const params = new HttpParams()
      .set('pageNumber', page.toString())
      .set('pageSize', size.toString());
    
    return this.http.get<UserPageResponse>(`${this.apiUrl}/active`, { params });
  }

  // Get user by ID
  getUserById(id: string): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/${id}`);
  }

  // Create new user
  createUser(userData: UserCreateRequest): Observable<User> {
    return this.http.post<User>(this.apiUrl, userData);
  }

  // Update user
  updateUser(id: string, userData: UserUpdateRequest): Observable<User> {
    return this.http.put<User>(`${this.apiUrl}/${id}`, userData);
  }

  // Delete user (deactivate)
  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
} 

File: services/cms/src/main/java/com/thmanyah/services/cms/services/UserService.java
package com.thmanyah.services.cms.services;

import com.thmanyah.services.cms.model.Role;
import com.thmanyah.services.cms.model.User;
import com.thmanyah.services.cms.model.dto.UserCreateDto;
import com.thmanyah.services.cms.model.dto.UserDto;
import com.thmanyah.services.cms.model.dto.UserUpdateDto;
import java.util.UUID;
import org.springframework.data.domain.Page;

public interface UserService {

  /**
   * Create a new user
   *
   * @param creatorEmail Email of the user creating this user
   * @param userCreateDto User creation data
   * @return Created user DTO
   */
  UserDto createUser(String creatorEmail, UserCreateDto userCreateDto);

  /**
   * Update an existing user
   *
   * @param updaterEmail Email of the user updating this user
   * @param id User ID
   * @param userUpdateDto Updated user data
   * @return Updated user DTO
   */
  UserDto updateUser(String updaterEmail, UUID id, UserUpdateDto userUpdateDto);

  /**
   * Get all users with pagination
   *
   * @param pageNumber Page number (0-based)
   * @param pageSize Page size
   * @return Page of user DTOs
   */
  Page<UserDto> getAllUsers(int pageNumber, int pageSize);

  /**
   * Get users by role with pagination
   *
   * @param role User role
   * @param pageNumber Page number (0-based)
   * @param pageSize Page size
   * @return Page of user DTOs filtered by role
   */
  Page<UserDto> getUsersByRole(Role role, int pageNumber, int pageSize);

  /**
   * Get active users with pagination
   *
   * @param pageNumber Page number (0-based)
   * @param pageSize Page size
   * @return Page of active user DTOs
   */
  Page<UserDto> getActiveUsers(int pageNumber, int pageSize);

  /**
   * Get user by ID as DTO
   *
   * @param id User ID
   * @return User DTO
   */
  UserDto getUserDtoById(UUID id);

  /**
   * Get user entity by ID
   *
   * @param id User ID
   * @return User entity
   */
  User getUserById(UUID id);

  /**
   * Get user by email
   *
   * @param email User email
   * @return User entity
   */
  User getUserByEmail(String email);

  /**
   * Delete a user (deactivate)
   *
   * @param deleterEmail Email of the user performing the deletion
   * @param id User ID
   */
  void deleteUser(String deleterEmail, UUID id);

  /**
   * Check if user exists by email
   *
   * @param email User email
   * @return true if exists, false otherwise
   */
  boolean existsByEmail(String email);

  /**
   * Authenticate user with email and password
   *
   * @param email User email
   * @param password Plain text password
   * @return User entity if authentication successful, null otherwise
   */
  User authenticateUser(String email, String password);

  /**
   * Check if user has admin role
   *
   * @param email User email
   * @return true if user is admin, false otherwise
   */
  boolean isAdmin(String email);

  /**
   * Check if user has editor role or higher
   *
   * @param email User email
   * @return true if user is editor or admin, false otherwise
   */
  boolean isEditorOrAdmin(String email);
}
